core/common/src/main/java/alluxio/metrics/MetricsSystem.java

```diff
@@ -412,6 +415,53 @@ public static void resetAllCounters() {
     }
   }
 
+  /**
+   * This method is used to return a list of RPC metric objects which will be sent to the
+   * MetricsMaster.
+   *
+   * It is mainly useful for metrics which have a {@link MetricType} of COUNTER. The metric values
+   * with this type will be only include value of the difference since the last time that
+   * {@code reportMetrics} was called.
+   *
+   * By sending only the difference since the last RPC, this method will allow the master to
+   * track the metrics for a given worker, even if the worker is restarted.
+   */
+  private static List<alluxio.grpc.Metric> reportMetrics(InstanceType instanceType) {
+    List<alluxio.grpc.Metric> rpcMetrics = new ArrayList<>(20);
+    for (Metric m : allMetrics(instanceType)) {
+      // last reported metrics only need to be tracked for COUNTER metrics
+      // Store the last metric value which was sent, but the rpc metric returned should only
+      // contain the difference of the current value, and the last value sent. If it doesn't
+      // yet exist, just send the current value
+      if (m.getMetricType() == MetricType.COUNTER) {
+        Metric prev = LAST_REPORTED_METRICS.replace(m.getFullMetricName(), m);
+        // On restarts counters will be reset to 0, so whatever the value is the first time
+        // this method is called represents the value which should be added to the master's
+        // counter.
+        double diff = prev != null ? m.getValue() - prev.getValue() : m.getValue();
+        rpcMetrics.add(new Metric(m.getInstanceType(), m.getHostname(), m.getInstanceId(),
+            m.getMetricType(), m.getName(), diff).toProto());
```

Comment:
Could do `toProto`, `toBuilder` and modify the value?

Reply:
Fixed